src/canvas/Canvas.ts

```diff
@@ -1096,12 +1107,12 @@ export class Canvas extends SelectableCanvas {
     const pointer = this.getPointer(e, true);
     // save pointer for check in __onMouseUp event
     this._previousPointer = pointer;
-    const shouldRender = this._shouldRender(target),
-      shouldGroup = this._shouldGroup(e, target);
+    const shouldRender = this._shouldRender(target);
+    let shouldGroup = false;
```

Comment:
how do we know that this was always false in all possible combinations?
If this is a micro optimization ( don't run should group if i should clear selection ) i would not do it.
or i would carefully check the commit history before changing it.

Reply:
I looked into this very closely.
Let me explain.

    shouldGroup = this._shouldGroup(e, target);
    if (this._shouldClearSelection(e, target)) {
      this.discardActiveObject(e);
    } else if (shouldGroup) {
      // in order for shouldGroup to be true, target needs to be true
      this._handleGrouping(e, target!);
      target = this._activeObject;
    }

`_shouldClearSelection` and `_shouldGroup` or opposites of each other

https://github.com/fabricjs/fabric.js/blob/84b7a357cef0c899ecd6b3fc0a74de2d57283d21/src/canvas/SelectableCanvas.ts#L725
https://github.com/fabricjs/fabric.js/blob/a39e01cf5ceae66809272051f77319345d8492d2/src/canvas/Canvas.ts#L1477

The whole point of this change is to squash them together. The logic is too complex and scattered around everywhere

Reply:
so when you agree I will squash them here

Reply:
https://github.com/fabricjs/fabric.js/commit/aaa5d1e84681491cbd196177f388eacfb65b57a0